File: services/shared/schemas/role.py
"""Role schema — Python mirror of role.ts."""

from __future__ import annotations

import re
from datetime import datetime

from pydantic import BaseModel, Field, field_validator

DEFAULT_ROLE_SLUGS: tuple[str, ...] = ("owner", "admin", "member", "client_guest")
_SLUG_RE = re.compile(r"^[a-z0-9_]+$")


class Role(BaseModel):
    id: str
    company_id: str
    slug: str = Field(..., min_length=1, max_length=60)
    display_name: str = Field(..., min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=2000)
    is_default: bool = False
    created_at: datetime

    @field_validator("slug")
    @classmethod
    def _slug_format(cls, v: str) -> str:
        if not _SLUG_RE.match(v):
            raise ValueError("slug must match ^[a-z0-9_]+$")
        return v


class RoleCreate(BaseModel):
    company_id: str
    slug: str = Field(..., min_length=1, max_length=60)
    display_name: str = Field(..., min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=2000)

    @field_validator("slug")
    @classmethod
    def _slug_format(cls, v: str) -> str:
        if not _SLUG_RE.match(v):
            raise ValueError("slug must match ^[a-z0-9_]+$")
        return v

File: services/shared/schemas/role.ts
import { z } from "zod";

// Workspace-scoped RBAC role definitions.
// Default roles are seeded per company in `db/migrations/0003_rbac_seed.sql`.
// Workspaces can clone defaults and add custom roles.

export const RoleSlugSchema = z.string().min(1).max(60).regex(/^[a-z0-9_]+$/);
export type RoleSlug = z.infer<typeof RoleSlugSchema>;

// The four defaults seeded for every workspace. Custom roles use any
// other slug matching the regex.
export const DEFAULT_ROLE_SLUGS = ["owner", "admin", "member", "client_guest"] as const;
export type DefaultRoleSlug = (typeof DEFAULT_ROLE_SLUGS)[number];

export const RoleSchema = z.object({
  id: z.string().uuid(),
  companyId: z.string().uuid(),
  slug: RoleSlugSchema,
  displayName: z.string().min(1).max(120),
  description: z.string().max(2000).nullable().optional(),
  isDefault: z.boolean().default(false),
  createdAt: z.string().datetime(),
});
export type Role = z.infer<typeof RoleSchema>;

export const RoleCreateSchema = RoleSchema.pick({
  companyId: true,
  slug: true,
  displayName: true,
  description: true,
});
export type RoleCreate = z.infer<typeof RoleCreateSchema>;
